### ISSUE TITLE
Problem in \StringUtil::parseSimpleTokens with hyphens

### ISSUE THREAD

**Issue by fritzmg** [contao/core#8436]
This issue arises in combination with [terminal42/contao-leads](https://github.com/terminal42/contao-leads) for example. Suppose you created a form in the form generator with a field named `E-mail`. Now suppose you also want to create a lead with that and you use `##E-mail##` in the __Record label__ setting, so that this field is shown in the backend, when viewing the entries of that lead.

However, if you try to do that, you won't be able to view that lead in the backend, because there will be a script timeout or a memory overflow. The `\Contao\StringUtil::parseSimpleTokens` function does not work if you have a simple token with a hyphen (`-`) in it.

So currently you simply have to either avoid using field names with hyphens or avoid using field names with hyphens in the __Record label__ - however this is something that won't be clear to the regular user.

@Toflar suggested that may be `\StringUtil::parseSimpleTokens` should convert invalid chars first before running
```php
preg_replace('/##([A-Za-z0-9_]+)##/i', '<?php echo $arrData[\'$1\']; ?>', $strReturn);
```

**Comment by Toflar** [comment]
@leofeyer You can assign this to me, I'll fix it.

**Comment by fritzmg** [comment]
Ok, the problem isn't actually in the Contao core. The function simply ignores such simple tokens, however, this causes an endless recursion in `\Haste\Util\StringUtil::recursiveReplaceTokensAndTags` - and that's the problem.

Sort of related to this problem: #8403 (i.e., if hyphens in field names would not be allowed, this problem would not occurs - however this can't be the solution in this case, I guess).

**Comment by Toflar** [comment]
This needs to be marked up for discussion anyway because I'm not happy with the behaviour of Contao at all.

**Comment by aschempp** [comment]
The solution should be fairly simple:

```php
$strReturn = str_replace(
    array_map(
        function ($v) {
            return '##' . $v . '##';
        },
        array_keys($arrData)
    ),
    $arrData,
    $strReturn
);

preg_replace('/##([A-Za-z0-9_]+)##/i', '', $strReturn);
```

**Comment by aschempp** [comment]
next version:

```php
$stReturn = preg_replace_callback(
    '/##([^#]+)##/i',
    function (array $matches) use ($arrData) {
        if (!isset($arrData[$matches[1]]) {
            System::log(sprintf('Unknown simple token "%s"', $matches[1]));
            return '';
        }

        return $arrData[$matches[1]];
    },
    $strReturn
);
```

**Comment by Toflar** [comment]
Fixed missing support for any character and added log entries: https://github.com/contao/core/pull/8438.
See screenshot for error messages :)

**Comment by leofeyer** [comment]
See #8438.

**Comment by leofeyer** [comment]
Fixed in a8d10c91e8e13dd4760277a35e3c99a319a1907c thanks to @Toflar and @ausi.
